src/plugins/intel_gna/src/scale_factor_helper.cpp

```diff
@@ -0,0 +1,117 @@
+        ::log::warning() << "Using input scale factor: " << input_scale_factors[id]
+                         << ", defined in configuration for input id: " << id << std::endl;
+        inputs.Get().at(id).scale_factor = input_scale_factors[id];
+    }
+}
+
+static bool IsCustomInputScaleFactorPerInputAvailable(const std::map<std::string, float>& per_input_scale_factors) {
+    if (per_input_scale_factors.empty()) {
+        return false;
+    }
+
+    bool is_scale_factor_custom = false;
+    for (const auto& scale_factor : per_input_scale_factors) {
+        if (!fp32eq(scale_factor.second, GNAPluginNS::kScaleFactorDefault)) {
+            is_scale_factor_custom = true;
+            break;
+        }
+    }
+
+    return is_scale_factor_custom;
+}
+
+static void ApplyScaleFactorsPerInput(const std::map<std::string, float>& per_input_scale_factors,
+                                      GNAPluginNS::GnaInputs& inputs) {
+    if (per_input_scale_factors.size() > inputs.size()) {
+        IE_THROW() << "Configuration per input scale factors count is bigger than inputs count";
+    }
+
+    for (auto&& sf : per_input_scale_factors) {
+        auto input_it = inputs.find(sf.first);
+        if (input_it == inputs.end()) {
+            IE_THROW() << "Given scale factor for invalid input: " << sf.first;
+        }
+        ::log::warning() << "Using input scale factor: " << sf.second
+                         << ", defined in configuration for input: " << sf.first << std::endl;
+        input_it->scale_factor = sf.second;
+    }
+}
+
+static bool CheckIfCanApplyCustomScaleFactor(const GNAPluginNS::HeaderLatest::ModelHeader& header) {
+    static constexpr GNAPluginNS::Header2dot8::ModelHeader::Version sc_forbid_overite_scale_factor;
+    if (!GNAPluginNS::HeaderLatest::IsFirstVersionLower(header.version, sc_forbid_overite_scale_factor)) {
+        ::log::warning() << "Cannot apply custom scale factor for model versions >= "
+                         << sc_forbid_overite_scale_factor.major << "." << sc_forbid_overite_scale_factor.minor
+                         << std::endl;
+        return false;
+    }
+    return true;
+}
+
+void ApplyInputScaleFactors(const GNAPluginNS::Config& config,
+                            const GNAPluginNS::HeaderLatest::ModelHeader& header,
+                            GNAPluginNS::GnaInputs& inputs) {
+    // If scale factors are defined in configuration we still need to use them instead of imported values,
+    // for example to change the scale factors for the old models.
+    const bool custom_scale_factor_per_input =
+        IsCustomInputScaleFactorPerInputAvailable(config.inputScaleFactorsPerInput);
+    const bool custom_scale_factor_legacy = IsCustomInputScaleFactorAvailableLegacy(config.inputScaleFactors);
+
+    if (!custom_scale_factor_per_input && !custom_scale_factor_legacy) {
+        return;
+    }
+
+    if (!CheckIfCanApplyCustomScaleFactor(header)) {
```

Comment:
It would better to apply ApplyScaleFactorsLegacy under this condition only:
if (CheckIfCanApplyCustomScaleFactor(header)) {
          ApplyScaleFactorsLegacy(config.inputScaleFactors, inputs);
}

Reply:
wanted to avoid double testing, and  inputScaleFactorsPerInput  should be checked applied as first if both are given.